src/librbd/Journal.cc

```diff
@@ -23,6 +23,17 @@
 
 namespace librbd {
 
+template<typename T>
```

Comment:
Minor: how about making this a member class and statically allocating it within ```librbd::Journal```?  That avoids the need to expose ```handle_metadata_updated``` as a public method.